### PR TITLE
Add a `deploy_env` attribute to `java_export`

### PR DESCRIPTION
**Pull request by shs96c**
This allows `java_export` to be used to generate things such as
plugins for Apache Spark or jar files for use in JEE servers by
excluding provided dependencies from the generated jars. This is
analagous to maven's `provided` scope, and maps to the `deploy_env`
attribute of `java_binary`.

The difference between `java_export` and `java_binary`'s `deploy_env`
is that `java_export` allows anything that provides a `JavaInfo` as an
input to `deploy_env`. This is because the expected use-cases are for
things that are normally provided as regular imports from a maven
dependency, rather than a custom binary.

### CODE REVIEW

File: docs/api.md [hunk 1]
```diff
@@ -1,18 +1,3 @@
-<!-- START doctoc generated TOC please keep comment here to allow auto update -->
```
Comment (jin) [inline docs/api.md:1]: Was this ran with https://github.com/bazelbuild/rules_jvm_external/blob/master/scripts/generate_docs.sh?

File: tests/integration/java_export/check-deploy-env.sh [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+  source "$(grep -sm1 "^$f " "$0.exe.runfiles_manifest" | cut -f2- -d' ')" 2>/dev/null || \
+  { echo>&2 "ERROR: cannot find $f"; exit 1; }; f=; set -e
+# --- end runfiles.bash initialization v2 ---
+
+set -euox pipefail
+
+classes_file=$(rlocation rules_jvm_external/tests/integration/java_export/classes.txt)
+
+if grep -q Dependency.class "$classes_file"; then
+  exit 1
```
Comment (jin) [inline tests/integration/java_export/check-deploy-env.sh:17]: echo an error message here?

File: tests/integration/java_export/check-deploy-env.sh [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+fi
+
+if ! grep -q Main.class "$classes_file"; then
+  exit 2
```
Comment (jin) [inline tests/integration/java_export/check-deploy-env.sh:21]: Why are the exit codes different?

**Review by jin — APPROVED** [review]

File: tests/integration/java_export/check-deploy-env.sh (hunk 2 above)
Comment (shs96c) [reply tests/integration/java_export/check-deploy-env.sh:18]: Done

File: tests/integration/java_export/check-deploy-env.sh (hunk 3 above)
Comment (shs96c) [reply tests/integration/java_export/check-deploy-env.sh]: To differentiate why the test failed. Echoing a message instead.

File: docs/api.md (hunk 1 above)
Comment (shs96c) [reply docs/api.md]: It is now :) Good catch. Thank you!

**Pull request merged**